### ISSUE TITLE
rds download-db-log-file-portion error: TypeError: 'NoneType' object has no attribute '__getitem__'

### ISSUE THREAD

**Issue by snarfed** [aws/aws-cli#1365]
`aws rds download-db-log-file-portion` currently fails for me on some of our log files, but not others. example debug log and stack trace below. the log files that fail mostly end in `-15` through `-01`, ie they're the hourly logs for 3pm through 1am, but only for some days. any ideas?

```
hell:~> aws --version
aws --version
aws-cli/1.7.31 Python/2.7.6 Darwin/14.3.0

hell:~> aws --debug --profile=prod --region=us-east-1 rds download-db-log-file-portion --db-instance-identifier default-prod --max-items 99999999 --starting-token 0 --output text --log-file-name error/postgresql.log.2015-05-29-20
aws --debug --profile=prod --region=us-east-1 rds download-db-log-file-portion --db-instance-identifier default-prod --max-items 99999999 --starting-token 0 --output text --log-file-name error/postgresql.log.2015-05-29-20
2015-06-02 20:46:50,823 - MainThread - awscli.clidriver - DEBUG - CLI version: aws-cli/1.7.31 Python/2.7.6 Darwin/14.3.0, botocore version: 1.0.0a3
2015-06-02 20:46:50,823 - MainThread - awscli.clidriver - DEBUG - Arguments entered to CLI: ['--debug', '--profile=prod', '--region=us-east-1', 'rds', 'download-db-log-file-portion', '--db-instance-identifier', 'default-prod', '--max-items', '99999999', '--starting-token', '0', '--output', 'text', '--log-file-name', 'error/postgresql.log.2015-05-29-20']
2015-06-02 20:46:50,823 - MainThread - botocore.hooks - DEBUG - Event session-initialized: calling handler <function add_scalar_parsers at 0x10c4925f0>
2015-06-02 20:46:50,824 - MainThread - botocore.hooks - DEBUG - Event session-initialized: calling handler <function inject_assume_role_provider at 0x10c446410>
2015-06-02 20:46:50,824 - MainThread - botocore.credentials - DEBUG - Skipping environment variable credential check because profile name was explicitly set.
2015-06-02 20:46:50,848 - MainThread - botocore.hooks - DEBUG - Event service-data-loaded.rds: calling handler <function register_retries_for_service at 0x10c0bc2a8>
2015-06-02 20:46:50,849 - MainThread - botocore.handlers - DEBUG - Registering retry handlers for service: rds
2015-06-02 20:46:50,851 - MainThread - botocore.hooks - DEBUG - Event building-command-table.rds: calling handler <function _building_command_table at 0x10c3cf320>
2015-06-02 20:46:50,851 - MainThread - botocore.hooks - DEBUG - Event service-data-loaded.rds: calling handler <function register_retries_for_service at 0x10c0bc2a8>
2015-06-02 20:46:50,852 - MainThread - botocore.handlers - DEBUG - Registering retry handlers for service: rds
2015-06-02 20:46:50,887 - MainThread - botocore.hooks - DEBUG - Event building-command-table.rds: calling handler <function add_waiters at 0x10c44b500>
2015-06-02 20:46:50,890 - MainThread - awscli.clidriver - DEBUG - OrderedDict([(u'db-instance-identifier', <awscli.arguments.CLIArgument object at 0x10c6fc850>), (u'log-file-name', <awscli.arguments.CLIArgument object at 0x10c6fc550>), (u'marker', <awscli.arguments.CLIArgument object at 0x10c6fc890>), (u'number-of-lines', <awscli.arguments.CLIArgument object at 0x10c6fc8d0>)])
2015-06-02 20:46:50,890 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.rds.download-db-log-file-portion: calling handler <function add_streaming_output_arg at 0x10c32ded8>
2015-06-02 20:46:50,890 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.rds.download-db-log-file-portion: calling handler <function add_cli_input_json at 0x10c43a7d0>
2015-06-02 20:46:50,890 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.rds.download-db-log-file-portion: calling handler <function unify_paging_params at 0x10c33ff50>
2015-06-02 20:46:50,972 - MainThread - awscli.customizations.paginate - DEBUG - Modifying paging parameters for operation: DownloadDBLogFilePortion
2015-06-02 20:46:50,972 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.rds.download-db-log-file-portion: calling handler <function add_generate_skeleton at 0x10c43aaa0>
2015-06-02 20:46:50,973 - MainThread - botocore.hooks - DEBUG - Event before-building-argument-table-parser.rds.download-db-log-file-portion: calling handler <bound method CliInputJSONArgument.override_required_args of <awscli.customizations.cliinputjson.CliInputJSONArgument object at 0x10c6fc910>>
2015-06-02 20:46:50,973 - MainThread - botocore.hooks - DEBUG - Event before-building-argument-table-parser.rds.download-db-log-file-portion: calling handler <bound method GenerateCliSkeletonArgument.override_required_args of <awscli.customizations.generatecliskeleton.GenerateCliSkeletonArgument object at 0x10c6fcc90>>
2015-06-02 20:46:51,037 - MainThread - botocore.hooks - DEBUG - Event operation-args-parsed.rds.download-db-log-file-portion: calling handler <functools.partial object at 0x10c618d08>
2015-06-02 20:46:51,037 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.db-instance-identifier: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,037 - MainThread - botocore.hooks - DEBUG - Event process-cli-arg.rds.download-db-log-file-portion: calling handler <awscli.argprocess.ParamShorthand object at 0x10c498290>
2015-06-02 20:46:51,037 - MainThread - awscli.argprocess - DEBUG - Detected structure: scalar
2015-06-02 20:46:51,037 - MainThread - awscli.arguments - DEBUG - Unpacked value of u'default-prod' for parameter "db_instance_identifier": u'default-prod'
2015-06-02 20:46:51,037 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.log-file-name: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,106 - MainThread - botocore.hooks - DEBUG - Event process-cli-arg.rds.download-db-log-file-portion: calling handler <awscli.argprocess.ParamShorthand object at 0x10c498290>
2015-06-02 20:46:51,106 - MainThread - awscli.argprocess - DEBUG - Detected structure: scalar
2015-06-02 20:46:51,106 - MainThread - awscli.arguments - DEBUG - Unpacked value of u'error/postgresql.log.2015-05-29-20' for parameter "log_file_name": u'error/postgresql.log.2015-05-29-20'
2015-06-02 20:46:51,107 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.marker: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,107 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.number-of-lines: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,107 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.cli-input-json: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,210 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.starting-token: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,210 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.page-size: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,210 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.max-items: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,210 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.rds.download-db-log-file-portion.generate-cli-skeleton: calling handler <function uri_param at 0x10c3210c8>
2015-06-02 20:46:51,210 - MainThread - botocore.hooks - DEBUG - Event calling-command.rds.download-db-log-file-portion: calling handler <bound method GenerateCliSkeletonArgument.generate_json_skeleton of <awscli.customizations.generatecliskeleton.GenerateCliSkeletonArgument object at 0x10c6fcc90>>
2015-06-02 20:46:51,293 - MainThread - botocore.hooks - DEBUG - Event calling-command.rds.download-db-log-file-portion: calling handler <bound method CliInputJSONArgument.add_to_call_parameters of <awscli.customizations.cliinputjson.CliInputJSONArgument object at 0x10c6fc910>>
2015-06-02 20:46:51,293 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: assume-role
2015-06-02 20:46:51,293 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: shared-credentials-file
2015-06-02 20:46:51,294 - MainThread - botocore.credentials - INFO - Found credentials in shared credentials file: ~/.aws/credentials
2015-06-02 20:46:51,295 - MainThread - botocore.client - DEBUG - Registering retry handlers for service: rds
2015-06-02 20:46:51,297 - MainThread - botocore.hooks - DEBUG - Event creating-client-class.rds: calling handler <function add_generate_presigned_url at 0x10c095f50>
2015-06-02 20:46:51,307 - MainThread - botocore.endpoint - DEBUG - Making request for <botocore.model.OperationModel object at 0x10c724410> (verify_ssl=True) with params: {'url_path': '/', 'body': {u'LogFileName': u'error/postgresql.log.2015-05-29-20', 'Version': u'2014-10-31', u'DBInstanceIdentifier': u'default-prod', u'Marker': u'0', 'Action': u'DownloadDBLogFilePortion'}, 'headers': {'User-Agent': 'aws-cli/1.7.31 Python/2.7.6 Darwin/14.3.0'}, 'url': u'https://rds.amazonaws.com/', 'query_string': '', 'method': u'POST'}
2015-06-02 20:46:51,372 - MainThread - botocore.hooks - DEBUG - Event request-created.rds.DownloadDBLogFilePortion: calling handler <bound method rds._sign_request of <botocore.client.rds object at 0x10c707e10>>
2015-06-02 20:46:51,373 - MainThread - botocore.auth - DEBUG - Calculating signature using v4 auth.
2015-06-02 20:46:51,373 - MainThread - botocore.auth - DEBUG - CanonicalRequest:
POST
/

host:rds.amazonaws.com
user-agent:aws-cli/1.7.31 Python/2.7.6 Darwin/14.3.0
x-amz-date:20150603T034651Z

host;user-agent;x-amz-date
XXX
2015-06-02 20:46:51,443 - MainThread - botocore.auth - DEBUG - StringToSign:
AWS4-HMAC-SHA256
20150603T034651Z
20150603/us-east-1/rds/aws4_request
XXX
2015-06-02 20:46:51,443 - MainThread - botocore.auth - DEBUG - Signature:
XXX
2015-06-02 20:46:51,449 - MainThread - botocore.endpoint - DEBUG - Sending http request: <PreparedRequest [POST]>
2015-06-02 20:46:51,449 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - INFO - Starting new HTTPS connection (1): rds.amazonaws.com
2015-06-02 20:46:52,846 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - DEBUG - "POST / HTTP/1.1" 200 355
2015-06-02 20:46:52,847 - MainThread - botocore.parsers - DEBUG - Response headers:
{'content-length': '355',
 'content-type': 'text/xml',
 'date': 'Wed, 03 Jun 2015 03:46:52 GMT',
 'x-amzn-requestid': 'XXX'}
2015-06-02 20:46:52,847 - MainThread - botocore.parsers - DEBUG - Response body:
<DownloadDBLogFilePortionResponse xmlns="http://rds.amazonaws.com/doc/2014-10-31/">
  <DownloadDBLogFilePortionResult>
    <AdditionalDataPending>false</AdditionalDataPending>
  </DownloadDBLogFilePortionResult>
  <ResponseMetadata>
    <RequestId>XXX</RequestId>
  </ResponseMetadata>
</DownloadDBLogFilePortionResponse>

2015-06-02 20:46:52,848 - MainThread - botocore.hooks - DEBUG - Event needs-retry.rds.DownloadDBLogFilePortion: calling handler <botocore.retryhandler.RetryHandler object at 0x10c689f10>
2015-06-02 20:46:52,848 - MainThread - botocore.retryhandler - DEBUG - No retry needed.
2015-06-02 20:46:52,848 - MainThread - botocore.hooks - DEBUG - Event after-call.rds.DownloadDBLogFilePortion: calling handler <awscli.errorhandler.ErrorHandler object at 0x10c498590>
2015-06-02 20:46:52,940 - MainThread - awscli.errorhandler - DEBUG - HTTP Response Code: 200
2015-06-02 20:46:52,941 - MainThread - awscli.clidriver - DEBUG - Exception caught in main()
Traceback (most recent call last):
  File "/usr/local/Cellar/awscli/1.7.31/libexec/lib/python2.7/site-packages/awscli/clidriver.py", line 183, in main
    return command_table[parsed_args.command](remaining, parsed_args)
  File "/usr/local/Cellar/awscli/1.7.31/libexec/lib/python2.7/site-packages/awscli/clidriver.py", line 369, in __call__
    return command_table[parsed_args.operation](remaining, parsed_globals)
  File "/usr/local/Cellar/awscli/1.7.31/libexec/lib/python2.7/site-packages/awscli/clidriver.py", line 538, in __call__
    call_parameters, parsed_globals)
  File "/usr/local/Cellar/awscli/1.7.31/libexec/lib/python2.7/site-packages/awscli/clidriver.py", line 663, in invoke
    self._display_response(operation_name, response, parsed_globals)
  File "/usr/local/Cellar/awscli/1.7.31/libexec/lib/python2.7/site-packages/awscli/clidriver.py", line 672, in _display_response
    formatter(command_name, response)
  File "/usr/local/Cellar/awscli/1.7.31/libexec/lib/python2.7/site-packages/awscli/formatter.py", line 234, in __call__
    for page in response:
  File "/usr/local/Cellar/awscli/1.7.31/libexec/vendor/lib/python2.7/site-packages/botocore/paginate.py", line 91, in __iter__
    parsed, primary_result_key, starting_truncation)
  File "/usr/local/Cellar/awscli/1.7.31/libexec/vendor/lib/python2.7/site-packages/botocore/paginate.py", line 171, in _handle_first_request
    all_data[starting_truncation:]
TypeError: 'NoneType' object has no attribute '__getitem__'
2015-06-02 20:46:53,039 - MainThread - awscli.clidriver - DEBUG - Exiting with rc 255

'NoneType' object has no attribute '__getitem__'
```

**Comment by snarfed** [comment]
i'm not seeing these errors today, so i'm guessing the root cause was an AWS API error that awscli didn't handle well. the appropriate fix here is probably just better error handling for that case, whatever it is.

**Comment by kyleknap** [comment]
Thanks for the debug log. We will look into what caused the issue in the CLI. Let us know if you hit the issue again or if you gather any more information.

**Comment by rayluo** [comment]
Fixed by boto/botocore#700.

By the way, you do NOT need `--max-items 99999999` to call this function. You can find [some examples](https://github.com/aws/aws-cli/pull/1617/files) by `aws rds download-log-file-portion help`, too.
